packages/aws-cdk-lib/aws-stepfunctions/lib/states/distributed-map.ts

```diff
@@ -0,0 +1,770 @@
+   *
+   * @default - Distribute Map state will iterate over all items provided by the ItemReader
+   */
+  readonly maxItems?: number;
+
+  /**
+   * Render the ItemReader as JSON object
+   */
+  render(): any;
+
+  /**
+   * Compile policy statements to provide relevent permissions to the state machine
+   */
+  providePolicyStatements(): iam.PolicyStatement[];
+}
+
+/**
+ * Base interface for Item Reader configuration properties
+ */
+interface ItemReaderProps {
+  /**
+   * S3 Bucket containing objects to iterate over or a file with a list to iterate over
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * Limits the number of items passed to the Distributed Map state
+   *
+   * @default - Distribute Map state will iterate over all items provided by the ItemReader
+   */
+  readonly maxItems?: number;
+}
+
+/**
+ * Properties for configuring an Item Reader that iterates over objects in an S3 bucket
+ */
+export interface S3ObjectsItemReaderProps extends ItemReaderProps {
+  /**
+   * S3 prefix used to limit objects to iterate over
+   *
+   * @default - No prefix
+   */
+  readonly prefix?: string;
+}
+
+/**
+ * Item Reader configuration for iterating over objects in an S3 bucket
+ */
+export class S3ObjectsItemReader implements IItemReader {
+  /**
+   * S3 Bucket containing objects to iterate over
+   */
+  readonly bucket: IBucket;
+
+  /**
+   * S3 prefix used to limit objects to iterate over
+   *
+   * @default - No prefix
+   */
+  readonly prefix?: string;
+
+  /**
+   * Limits the number of items passed to the Distributed Map state
+   *
+   * @default - No maxItems
+   */
+  readonly maxItems?: number;
+
+  /**
+   * ARN for the `listObjectsV2` method of the S3 API
+   * This API method is used to iterate all objects in the S3 bucket/prefix
+   */
+  private readonly resource: string = `arn:${Aws.PARTITION}:states:::s3:listObjectsV2`;
+
+  constructor(props: S3ObjectsItemReaderProps) {
+    this.bucket = props.bucket;
+    this.prefix = props.prefix;
+    this.maxItems = props.maxItems;
+  }
+
+  /**
+   * Renders the ItemReader configuration as JSON object
+   * @returns - JSON object
+   */
+  public render(): any {
+    return FieldUtils.renderObject({
+      Resource: this.resource,
+      ...(this.maxItems && { ReaderConfig: { MaxItems: this.maxItems } }),
+      Parameters: {
+        Bucket: this.bucket.bucketName,
+        ...(this.prefix && { Prefix: this.prefix }),
+      },
+    });
+  }
+
+  /**
+   * Compile policy statements to provide relevent permissions to the state machine
+   */
+  public providePolicyStatements(): iam.PolicyStatement[] {
+    return [
+      new iam.PolicyStatement({
+        actions: [
+          's3:ListBucket',
+        ],
+        resources: [this.bucket.bucketArn],
+      }),
+    ];
+  }
+}
+
+/**
+ * Base interface for Item Reader configuration properties the iterate over entries in a S3 file
+ */
+export interface S3ItemReaderProps extends ItemReaderProps {
```

Comment:
Should this be `S3FileItemReader` or `S3FileReader`?

Reply:
I'll rename to `S3FileItemReader`